src/csharp/Tokenizer.cs

```diff
@@ -0,0 +1,133 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace Microsoft.ML.OnnxRuntimeGenAI
+{
+    public class Tokenizer : IDisposable
+    {
+        private IntPtr _tokenizerHandle;
+        private bool _disposed = false;
+
+        public Tokenizer(Model model)
+        {
+            Result.VerifySuccess(NativeMethods.OgaCreateTokenizer(model.Handle, out _tokenizerHandle));
+        }
+
+        public Sequences EncodeBatch(string[] strings)
+        {
+            IntPtr stringArray = IntPtr.Zero;
+            UIntPtr[] stringLengths = new UIntPtr[strings.Length];
+            for (int i = 0; i < strings.Length; i++)
+            {
+                stringLengths[i] = (UIntPtr)UTF8Encoding.UTF8.GetByteCount(strings[i]);
+            }
+
+            unsafe
+            {
+                fixed (UIntPtr* stringLengthsPtr = stringLengths)
+                {
+                    Result.VerifySuccess(NativeMethods.OgaCreateAllocatedStrings((UIntPtr)strings.Length, stringLengthsPtr, out stringArray));
+                }
+            }
+
+            try
+            {
+                for (ulong i = 0; i < (ulong)strings.Length; i++)
+                {
+                    Result.VerifySuccess(NativeMethods.OgaStringsGetBuffer(stringArray, (UIntPtr)i, out IntPtr buffer));
+                    Utils.ToNativeBuffer(strings[i], buffer, (int)stringLengths[i]);
+                }
+                Result.VerifySuccess(NativeMethods.OgaTokenizerEncodeBatchStrings(_tokenizerHandle, stringArray, out IntPtr nativeSequences));
+
+                return new Sequences(nativeSequences);
+            }
+            finally
+            {
+                NativeMethods.OgaDestroyStrings(stringArray);
+            }
+        }
+
+        public string[] DecodeBatch(Sequences sequences)
+        {
+            IntPtr stringArray = IntPtr.Zero;
+            Result.VerifySuccess(NativeMethods.OgaTokenizerDecodeBatchStrings(_tokenizerHandle, sequences.Handle, out stringArray));
```

Comment:
Same here, stringArray seems to be a pure overhead.

Reply:
Removed OgaStrings type. And will call Encode and Decode in a loop for the batch calls.